src/java.base/share/classes/java/security/Provider.java

```diff
@@ -913,149 +913,157 @@ private void implPutAll(Map<?,?> t) {
     }
 
     private Object implRemove(Object key) {
-        if (key instanceof String) {
-            if (!checkLegacy(key)) {
-                return null;
-            }
-            legacyStrings.remove((String)key);
+        if (!checkLegacy(key)) return null;
+
+        Object o = super.remove(key);
+        if (o != null && o instanceof String && key instanceof String) {
+            parseLegacy((String)key, (String)o, OPType.REMOVE);
         }
-        return super.remove(key);
+        return o;
     }
 
     private boolean implRemove(Object key, Object value) {
-        if (key instanceof String && value instanceof String) {
-            if (!checkLegacy(key)) {
-                return false;
-            }
-            legacyStrings.remove((String)key, (String)value);
+        if (!checkLegacy(key)) return false;
+
+        boolean result = super.remove(key, value);
+        if (result && key instanceof String && value instanceof String) {
+            parseLegacy((String)key, (String)value, OPType.REMOVE);
```

Comment:
I suggest to use pattern matching for instanceof to avoid manual casts

Reply:
Sure, will do so. Thanks for the suggestion.